### ISSUE TITLE
Design question

### ISSUE THREAD

**Issue by Ulrar** [andreas/ocaml-graphql-server#123]
Hi,

So I've been using the lib for a little while, and making simple stuff is fairly simple, which is great.
I do wonder how to make larger things the correct way though, since as far as I can tell there's no way to have a "mismatch" between the types the API exposes to the outside world and the types it's actually using to resolve the queries.

By that I mean if my schema has a field returning a list of objects, it has to actually return a list of objects in ocaml too which will then be passed to that object's resolve function to extract each field. That means all the way "up" in the root schema field, I have to resolve the whole object at once to be able to pass it down, even if the query is only trying to get a specific field for it.
For example in the schema : 

```ocaml
io_field "projects"
  ~typ:(non_null (list (non_null project)))
  ~args:Arg.[]
  ~resolve:
    (fun () () ->
          let%lwt res = Project.getAll blah in
          Lwt.return (Ok res)
    )
```

Since the type here is a list of project, which as far as I can tell I have to put there if I want that field to return a list of project, I end up having to resolve the whole list of projects here with every field for every project, whether the query actually wants a specific field or not. Since some of those fields can be more consuming than other to resolve, I end up having separate fields in my schema to allow querying different fields without forcing that resolve of everything at once.
For example, my "name" field is fairly cheap to resolve (I can get all the project names by listing a directory) but resolving other fields is more expensive because I have to parse files and decode a bunch of json. So when my frontend needs just the projects names, instead of querying "projects {name}" I end up having to add a separate field to my schema to make a different query entirely, something like "projectNames {}" which is a bit silly, I feel like I'm doing REST again.

I wonder if that's an actual limitation of the lib because of type checking, or if I'm missing something important here. Is there a way to "move" the resolution of the object further down, to say that I'll return a list of projects but without ever actually building that list ? My project Schema.obj has a resolve function for each field (which, like in the example, just does project.name for the name field for example), if those could take an arbitrary type instead of a project I could do the resolution there, which would be ideal.
So instead of having

```ocaml
obj "project"
  ~doc:"A project"
  ~fields:(fun _ ->
            [ field "name"
                ~doc:"Project name"
                ~typ:(non_null string)
                ~args:Arg.[]
                ~resolve:(fun _ p -> p.name)
            ; more fields
            ]
```
I'd love to be able to do the actual resolution for each field here, instead of having to take an already resolved project as an argument. For example if my field in the schema could return a list of strings (instead of a list of projects) I could pass down the list of projects names, and this resolve function for name would just be something like `~resolve:(fun _ name -> name)` but the resolve function for another field could do all the file access, json parsing and stuff needed to resolve that particular field only when the query actually calls for that field.

I hope I'm clear, I'm not sure, I've been using the word field a lot.
tl;dr I'm just looking for a way to avoid resolving unused fields when the query isn't asking for them.

Thanks !

**Comment by Ulrar** [comment]
Seems like it is indeed, from it's name I hadn't realized that was the subject of this other issue, sorry. 

Looks like the idea is to tell the "top" resolver what fields are being queried, which I imagine would allow it to fill the non-queried fields with garbage when building the objects instead of resolving them. I don't find that solution as nice as allowing the resolution to happen further down per field, but I imagine it would work, which is the important part. I'll take a serious look at it, thanks

**Comment by andreas** [comment]
Hi @Ulrar,

It's difficult to answer your question with certainty without more detail. It sounds like you have a schema something like the following:

```ocaml
let project = Schema.(obj "Project"
  ~fields:(fun _ -> [
    field "name"
      ~type:(non_null string)
      ~args:[]
      ~resolve:(fun _ctx project -> project.name)
    ;
    field "expensiveField"
      ~type:some_type
      ~args:[]
      ~resolve:(fun _ctx project -> (* the question is what to put here *))
  ])
)

let schema = Schema.(schema [
  field "projects"
    ~typ:(list projects)
    ~args:[]
    ~resolve:(fun _ctx () -> fetch_projects ())
])
```

If I understand correctly, your question is whether `expensiveField` needs to be loaded as part of `fetch_projects` or it can be done in the `expensiveField`-resolver.

The short answer is that the data can definitely be loaded in the `expensiveField`-resolver.

The slightly longer answer is that the most performant approach is application specific. Broadly speaking, I see 3 scenarios:

- *Scenario 1*: If `expensiveField` and `name` originate from the same data source, e.g. both are stored in MySQL, then the approach discussed in #42 is relevant. That is, `fetch_projects` need to load different data based on the query at hand. This scenario doesn't sound sound your case though, as you mention loading some data from disk.

- *Scenario 2*: If `expensiveField` benefits from batch loading, i.e. loading multiple values of `expensiveField` for many projects at once, then you can use the "data loader" approach. Check out [`facebook/dataloader`](https://github.com/facebook/dataloader) for a more elaborate description of this pattern. There's an early implementation of that in OCaml here: https://github.com/andreas/ocaml-dataloader.

- *Scenario 3*: If `expensiveField` does not benefit from batch loading, then you can simply load each value directly in the field resolver.

**Comment by Ulrar** [comment]
Hi,

Okay I get it, I thought `fetch_projects` had to return a list of projects, but after some testing I see I was wrong, you can return whatever you want as long as the project object takes that same type and constructs the correct object with that.
Makes a lot more sense now ! So I'll rework my thing with io_fields in the project resolvers instead, that's perfect, thanks !
